strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+
+    def squeeze(self, r, phi, k):
+        r"""Squeeze mode ``k`` by the amount ``r*exp(1j*phi)``.
+
+        Args:
+            r (float): squeezing magnitude
+            phi (float): squeezing phase
+            k (int): mode to be squeezed
+        """
+        if self.active[k] is None:
+            raise ValueError("Cannot squeeze mode, mode does not exist")
+
+        sq = symp.expand(symp.squeezing(r, phi), k, self.nlen)
+        self.means = update_means(self.means, sq, self.from_xp)
+        self.covs = update_covs(self.covs, sq, self.from_xp)
+
+    def mbsqueeze(self, k, r, phi, r_anc, eta_anc, avg):
+        r"""Squeeze mode ``k`` by the amount ``r*exp(1j*phi)`` using measurement-based squeezing.
+
+        Either the average map, described by a Gaussian CPTP transformation, or a single-shot map
+        with ancillary measurement outcomes can be simulated.
+
+        Args:
+            k (int): mode to be squeezed
+            r (float): target squeezing magnitude
+            phi (float): target squeezing phase
+            r_anc (float): squeezing magnitude of the ancillary mode
+            eta_anc(float): detection efficiency of the ancillary mode
+            avg (bool): whether to apply the average map or single-shot
+        Returns:
+            float: if single-shot map selected, returns the measurement outcome of the ancilla
+        """
+
+        if self.active[k] is None:
+            raise ValueError("Cannot squeeze mode, mode does not exist")
+
+        # antisqueezing corresponds to an extra phase shift
+        if r < 0:
+            phi += np.pi
+        r = np.abs(r)
+        # beamsplitter angle
+        theta = np.arccos(np.exp(-r))
+        self.phase_shift(-phi / 2, k)
+
+        # Construct (X,Y) for Gaussian CPTP of average map
+        if avg:
+            X = np.diag([np.cos(theta), 1 / np.cos(theta)])
+            Y = np.diag(
+                [
+                    (np.sin(theta) ** 2) * (np.exp(-2 * r_anc)),
+                    (np.tan(theta) ** 2) * (1 - eta_anc) / eta_anc,
+                ]
+            )
+            Y *= self.hbar / 2
+            X2, Y2 = self.expandXY([k], X, Y)
+            self.apply_channel(X2, Y2)
+
+        # Add new ancilla mode, interfere it and measure it
+        # Delete ancilla mode from active list
+        elif not avg:
+            self.add_mode()
+            new_mode = self.nlen - 1
+            self.squeeze(r_anc, 0, new_mode)
+            self.beamsplitter(theta, 0, k, new_mode)
+            self.loss(eta_anc, new_mode)
+            self.phase_shift(np.pi / 2, new_mode)
+            val = self.homodyne(new_mode)
+            self.del_mode(new_mode)
+            self.covs = np.delete(self.covs, [2 * new_mode, 2 * new_mode + 1], axis=1)
+            self.covs = np.delete(self.covs, [2 * new_mode, 2 * new_mode + 1], axis=2)
+            self.means = np.delete(self.means, [2 * new_mode, 2 * new_mode + 1], axis=1)
+            self.nlen = self.nlen - 1
```

Comment:
Would we not like to make this change in `del_mode`?
```suggestion
            self.nlen -= 1
```

Reply:
I think the spirit of del_mode is to delete modes that the user programs in, but the record that the mode was there is via self.nlen. Whereas here we really don't want the user to be able to access this ancillary mode, except for the measurement outcome from it.